src/Chapter3.hs

```diff
@@ -1006,6 +1145,108 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data Monster = Monster
+  { monsterName :: String
+  , monsterHealth :: Health
+  , monsterAttack :: Attack
+  , monsterActions :: [Action]
+  } deriving Show
+
+data Knight = Knight
+  { knightName :: String
+  , knightHealth :: Health
+  , knightAttack :: Attack
+  , knightDefense :: Defense
+  , knightActions :: [Action]
+  } deriving Show
+
+data Action
+  = MonsterAttack
+  | MonsterRun
+  | KnightAttack
+  | KnightHeal
+  | KnightCastSpell
+  deriving Show
```

Comment:
Ideally, you want to have separate data types for `MonsterAction` and `KnightAction`, so it won't be possible to mix `KnightAttack` and `MonsterRun` in a single list

Reply:
So I actually tried splitting it up but I kept getting errors about not being able to deduce the type for `Action` or `Fighter` from the instance of `getAction`, `updateAction`, and `takeAction`. I'm not sure how to correctly combine those two classes.

I pushed up my latest attempt.

Reply:
Scratch that, I figured it out. Wow, working through that really taught me a lot about the scope if instances. Again, really awesome exercises, and thank you sooo much for taking the time to review the PRs and provide super useful resources!

I'm pretty happy with the solution to this and the city problem, but the code feels a bit repetitive. Do you have any suggestions to help with that at this level, or is that something that I will need better abstractions for (via Chapter 4 and beyond)?